skbio/metadata/tests/test_interval.py

```diff
@@ -0,0 +1,505 @@
+        for value in [1, '', None]:
+            with self.assertRaises(TypeError):
+                f.metadata = value
+
+    def test_set_get_bad(self):
+        im = IntervalMetadata()
+        f = Interval(interval_metadata=im,
+                     locations=[(1, 2), (4, 7)],
+                     boundaries=[(True, False), (False, False)],
+                     metadata={'name': 'sagA', 'function': 'transport'})
+        f.drop()
+        with self.assertRaises(RuntimeError):
+            f.boundaries
+        with self.assertRaises(RuntimeError):
+            f.locations
+        with self.assertRaises(RuntimeError):
+            f.metadata
+        with self.assertRaises(RuntimeError):
+            f.boundaries = None
+        with self.assertRaises(RuntimeError):
+            f.locations = [(1, 2)]
+        with self.assertRaises(RuntimeError):
+            f.metadata = {}
+
+
+class TestIntervalUtil(unittest.TestCase):
+    def test_assert_valid_location(self):
+        intvls = [(1, 2), (-1, 2)]
+        for intvl in intvls:
+            try:
+                _assert_valid_location(intvl)
+            except:
+                self.assertTrue(False)
+
+    def test_assert_valid_location_wrong_type(self):
+        intvls = [[1, 2], 1, [1, 2, 3]]
+        for intvl in intvls:
+            with self.assertRaises(TypeError):
+                _assert_valid_location(intvl)
+
+    def test_assert_valid_location_wrong_value(self):
+        intvls = [(2, 1), (True, 0), ('s', 'r')]
+        for intvl in intvls:
+            with self.assertRaises(ValueError):
+                _assert_valid_location(intvl)
+
+    def test_assert_valid_boundary(self):
+        boundaries = [(True, False), (True, True)]
+        for boundary in boundaries:
+            try:
+                _assert_valid_boundary(boundary)
+            except:
+                self.assertTrue(False)
+
+    def test_assert_valid_boundary_wrong_value(self):
+        boundaries = [(True, False, True), (0, 1), ('s', '')]
+        for boundary in boundaries:
+            with self.assertRaises(ValueError):
+                _assert_valid_boundary(boundary)
+
+    def test_assert_valid_boundary_wrong_type(self):
+        boundaries = [[True, False], 's', 1]
+        for boundary in boundaries:
+            with self.assertRaises(TypeError):
+                _assert_valid_boundary(boundary)
+
+
+class TestIntervalMetadata(unittest.TestCase):
+    def setUp(self):
+        self.im_empty = IntervalMetadata()
+        self.im_1 = IntervalMetadata()
+        self.interval_1 = Interval(
+            interval_metadata=self.im_1,
+            locations=[(1, 2), (4, 7)],
+            metadata={'gene': 'sagA',  'location': 0})
+        self.im_2 = deepcopy(self.im_1)
+        self.interval_2 = Interval(
+            interval_metadata=self.im_2,
+            locations=[(3, 5)],
+            metadata={'gene': 'sagB', 'location': 0})
```

Comment:
There should be tests where the interval metadata has deeply equal `Interval` objects ensuring that methods such as `query` and `drop` understand them uniquely. e.g. query would still return two `Interval` objects even though they are deeply equal.